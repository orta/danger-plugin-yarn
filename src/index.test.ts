import * as mockfs from "fs"

import yarn, { checkForLockfileDiff, checkForNewDependencies, checkForRelease, checkForTypesInDeps } from "./index"

declare const global: any
beforeEach(() => {
  global.warn = jest.fn()
  global.message = jest.fn()
  global.fail = jest.fn()
  global.markdown = jest.fn()
  global.danger = { utils: { sentence: jest.fn() } }
})

afterEach(() => {
  global.warn = undefined
  global.message = undefined
  global.fail = undefined
  global.markdown = undefined
})

describe("checkForRelease", () => {
  it("Says congrats if there is a package diff version change", () => {
    checkForRelease({ version: {} })
    expect(global.markdown).toHaveBeenCalledWith(":tada:")
  })

  it("does nothing when there's no version change", () => {
    checkForRelease({})
    expect(global.markdown).toHaveBeenCalledTimes(0)
  })
})

describe("checkForTypesInDeps", () => {
  it("does nothing when there's no dependency changes", () => {
    checkForTypesInDeps({})
    expect(global.fail).toHaveBeenCalledTimes(0)
  })

  it("when there is an @types dependency, it should call fail", () => {
    const deps = {
      dependencies: {
        added: ["@types/danger"],
      },
    }
    checkForTypesInDeps(deps)
    expect(global.fail).toHaveBeenCalledTimes(1)
  })
})

describe("checkForLockfileDiff", () => {
  it("does nothing when there's no dependency changes", () => {
    checkForLockfileDiff({})
    expect(global.warn).toHaveBeenCalledTimes(0)
  })

  it("when there are dependency changes, and no lockfile in modified - warn", () => {
    global.danger = { git: { modified_files: [] } }
    const deps = {
      dependencies: {},
    }
    checkForLockfileDiff(deps)
    expect(global.warn).toHaveBeenCalledTimes(1)
  })

  it("when there are dependency changes, and a lockfile in modified - do not warn", () => {
    global.danger = { git: { modified_files: ["yarn.lock"] } }
    const deps = { dependencies: {} }
    checkForLockfileDiff(deps)
    expect(global.warn).toHaveBeenCalledTimes(0)
  })
})

describe("npm metadata", () => {
  it("Shows a bunch of useful text for a new dep", async () => {
    jest.mock("node-fetch", () => () =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(JSON.parse(mockfs.readFileSync("src/fixtures/danger-npm-info.json", "utf8"))),
      })
    )
    const { getNPMMetadataForDep } = require("./")
    const data = await getNPMMetadataForDep("danger")
    expect(data).toMatchSnapshot()
  })
  it("looks through versions if license is missing", async () => {
    jest.mock("node-fetch", () => () =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(JSON.parse(mockfs.readFileSync("src/fixtures/pinpoint-npm-info.json", "utf8"))),
      })
    )
    const { getNPMMetadataForDep } = require("./")
    const data = await getNPMMetadataForDep("pinpoint")
    expect(data).toMatchSnapshot()
  })
})
