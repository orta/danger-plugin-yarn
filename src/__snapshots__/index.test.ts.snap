// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm metadata Shows a bunch of useful text for a new dep 1`] = `
"
<h2><a href=\\"https://github.com/danger/danger-js#readme\\">danger</a></h2>
<p>Author: Orta Therox</p>
<p>Description: Unit tests for Team Culture</p>
<p>Homepage: <a href=\\"https://github.com/danger/danger-js#readme\\">https://github.com/danger/danger-js#readme</a></p>

<table>
  <thead><tr><th></th><th width=\\"100%\\"></th></tr></thead>
  <tr><td>Created</td><td>about 1 year ago</td></tr><tr><td>Last Updated</td><td>2 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>3</td></tr><tr><td>Releases</td><td>58</td></tr><tr><td>Direct Dependencies</td><td>undefined</td></tr><tr><td>Keywords</td><td>undefined</td></tr>
</table>

<details>
<summary><code>README</code></summary>
# Danger JS :no_entry_sign: [![npm](https://img.shields.io/npm/v/danger.svg)](https://www.npmjs.com/package/danger) [![Build Status](https://travis-ci.org/danger/danger-js.svg?branch=master)](https://travis-ci.org/danger/danger-js) [![Build Status](https://ci.appveyor.com/api/projects/status/ep5hgeox3lbc5c7f?svg=true)](https://ci.appveyor.com/project/orta/danger-js/branch/master)

Formalize your Pull Request etiquette.

<p align=\\"center\\">
    <a href=\\"#what-is-danger-js\\">What is Danger JS?</a> &bull;
    <a href=\\"VISION.md\\">Vision</a> &bull;
    <a href=\\"#this-thing-is-broken-i-should-help-improve-it\\">Helping Out</a> &bull;
    <a href=\\"http://danger.systems/js/usage/extending-danger.html\\">Plugin Development</a>
</p>

## What is Danger JS?

Danger runs after your CI, automating your team's conventions surrounding code review.

This provides another logical step in your process, through this Danger can help lint your rote tasks in daily code review.

You can use Danger to codify your teams norms, leaving humans to think about harder problems.

Danger JS currently works with GitHub and Travis CI, Circle CI, Semaphore, Jenkins, Docker Cloud, surf-build, Codeship, Drone, Buildkite, buddybuild or Visual Studio Team Services.

## For example?

You can:

-   Enforce CHANGELOGs
-   Enforce links to Trello/JIRA in PR/MR bodies
-   Enforce using descriptive labels
-   Look out for common anti-patterns
-   Highlight interesting build artifacts
-   Give warnings when specific files change

Danger provides the glue to let _you_ build out the rules specific to your team's culture, offering useful metadata and a comprehensive plugin system to share common issues.

## Getting Started

Alright. So, actually, you may be in the wrong place. From here on in, this README is going to be for people who are interested in working on and improving on Danger JS.

We keep all of the end-user documentation at <http://danger.systems/js>.

Some quick links to get you started:

-   [Getting Started](http://danger.systems/js/guides/getting_started.html) 
-   [Guides Index](http://danger.systems/js/guides.html)
-   [DSL Reference](http://danger.systems/js/reference.html)

## This thing is broken, I should help improve it!

Awesommmmee.

\`\`\`sh
git clone https://github.com/danger/danger-js.git
cd danger-js

# if you don't have yarn installed
npm install -g yarn

yarn install
\`\`\`

You can then verify your install by running the tests, and the linters:

\`\`\`sh
yarn test
yarn lint
\`\`\`

The fixers for both tslint and prettier will be applied when you commit, and on a push your code will be verified
that it compiles.

### What is the TODO?

Check the issues, I try and keep my short term perspective there. Long term is in the [VISION.md](VISION.md).

### Releasing a new version of Danger

Following [this commit](https://github.com/danger/danger-js/commit/a26ac3b3bd4f002acd37f6a363c8e74c9d5039ab) as a model:

-   Checkout the \`master\` branch. Ensure your working tree is clean, and make sure you have the latest changes by running \`git pull\`.
-   Update \`package.json\` with the new version - for the sake of this example, the new version is **0.21.0**.
-   Modify \`changelog.md\`, adding a new \`### 0.21.0\` heading under the \`### Master\` heading at the top of the file.
-   Commit both changes with the commit message **Version bump**.
-   Tag this commit - \`git tag 0.21.0\`.
-   Push the commit and tag to master - \`git push origin master --follow-tags\`. Travis CI will build the tagged commit and publish that tagged version to NPM.

:ship:

## License, Contributor's Guidelines and Code of Conduct

We try to keep as much discussion as possible in GitHub issues, but also have a pretty inactive Slack --- if you'd like an invite, ping [@Orta](https://twitter.com/orta/) a DM on Twitter with your email. It's mostly interesting if you want to stay on top of Danger without all the emails from GitHub.

> This project is open source under the MIT license, which means you have full access to the source code and can modify it to fit your own needs.
>
> This project subscribes to the [Moya Contributors Guidelines](https://github.com/Moya/contributors) which TLDR: means we give out push access easily and often.
>
> Contributors subscribe to the [Contributor Code of Conduct](http://contributor-covenant.org/version/1/3/0/) based on the [Contributor Covenant](http://contributor-covenant.org) version 1.3.0.

[emiss]: https://github.com/artsy/emission/blob/master/dangerfile.ts

[danger-js]: https://github.com/danger/danger-js/blob/master/dangerfile.ts

[meta]: https://github.com/artsy/metaphysics/blob/master/dangerfile.js

[fbj]: https://github.com/facebook/jest/blob/master/dangerfile.js

[sc]: https://github.com/styled-components/styled-components/blob/master/dangerfile.js

[rxjs]: https://github.com/ReactiveX/rxjs/blob/master/dangerfile.js

[setup]: http://danger.systems/guides/getting_started.html#creating-a-bot-account-for-danger-to-use

[jest]: https://github.com/facebook/jest

</details>

"
`;

exports[`npm metadata looks through versions if license is missing 1`] = `
"
<h2><a href=\\"http://npmjs.com/package/pinpoint\\">pinpoint</a></h2>
<p>Author: Glen Huang</p>
<p>Description: Display an arrow in a string of code to point to a location</p>
<p>Homepage: <a href=\\"http://npmjs.com/package/pinpoint\\">http://npmjs.com/package/pinpoint</a></p>

<table>
  <thead><tr><th></th><th width=\\"100%\\"></th></tr></thead>
  <tr><td>Created</td><td>almost 4 years ago</td></tr><tr><td>Last Updated</td><td>almost 4 years ago</td></tr><tr><td>License</td><td><b>MIT</b></td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>2</td></tr>
</table>

<details>
<summary><code>README</code></summary>
# pinpoint

Add line numbers and an arrow to a string of code that points to a specific location.

## Installation

	npm install pinpoint

## Example

With this code:

\`\`\`javascript
function add(left, right) {
  return left + right;
}

console.log(add(1 + 1));
\`\`\`

When being asked to point to line 2 and column 2, generates this:

\`\`\`javascript
1| function add(left, right) {
2|   return left + right;
-----^
3| }
4|
5| console.log(add(1 + 1));
\`\`\`
## API

\`\`\`javascript
var pinpoint = require('pinpoint');

var str = pinpoint(code, options);
\`\`\`

* \`code\` - a string of code
* \`options\` - an object literal supports these options:
	* \`line\` - the line number to point to, 1-based.
	* \`column\` - the column number to point to, 1-based.
	* \`showLines\` - the number of lines to show. The pointed line will try to stay in the middle of the displays lines.
	* \`indent\` - indent string prepend to each line, before the line number.
</details>

"
`;
